File: src/parser_utils/output/__tests__/layout_test.ml
(**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open OUnit2
open Layout
open Layout_test_utils

let tests =
  "layout"
  >::: [
         ( "fuse_list" >:: fun ctxt ->
           let a = Atom "a" in
           let b = Atom "b" in
           let c = Atom "c" in
           let d = Atom "d" in
           let sep = Atom "," in
           let actual = fuse_list [a; b] in
           let expected = Concat [a; pretty_space; b] in
           assert_layout ~ctxt expected actual;

           let actual = fuse_list [a; b; c] in
           let expected = Concat [a; pretty_space; b; pretty_space; c] in
           assert_layout ~ctxt expected actual;

           let actual = fuse_list ~sep [a; b] in
           let expected = Concat [a; sep; pretty_space; b] in
           assert_layout ~ctxt expected actual;

           let actual = fuse_list ~sep [a; b; c] in
           let expected = Concat [a; sep; pretty_space; b; sep; pretty_space; c] in
           assert_layout ~ctxt expected actual;

           let actual = fuse_list ~wrap:(c, d) [a; b] in
           let expected = Concat [c; a; pretty_space; b; d] in
           assert_layout ~ctxt expected actual;

           let actual = fuse_list ~sep ~wrap:(c, d) [a; b] in
           let expected = Concat [c; a; sep; pretty_space; b; d] in
           assert_layout ~ctxt expected actual );
       ]

File: rust_port/crates/flow_parser_utils_output/src/layout_test.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use crate::layout::LayoutNode;
use crate::layout::fuse_list;
use crate::layout_test_utils::assert_layout;
use crate::layout_test_utils::layout_builder::pretty_space;

#[test]
fn test_fuse_list() {
    let a = LayoutNode::atom("a".to_string());
    let b = LayoutNode::atom("b".to_string());
    let c = LayoutNode::atom("c".to_string());
    let d = LayoutNode::atom("d".to_string());
    let sep = LayoutNode::atom(",".to_string());

    let actual = fuse_list(None, None, vec![a.clone(), b.clone()]);
    let expected = LayoutNode::concat(vec![a.clone(), pretty_space(), b.clone()]);
    assert_layout(expected, actual);

    let actual = fuse_list(None, None, vec![a.clone(), b.clone(), c.clone()]);
    let expected = LayoutNode::concat(vec![
        a.clone(),
        pretty_space(),
        b.clone(),
        pretty_space(),
        c.clone(),
    ]);
    assert_layout(expected, actual);

    let actual = fuse_list(Some(sep.clone()), None, vec![a.clone(), b.clone()]);
    let expected = LayoutNode::concat(vec![a.clone(), sep.clone(), pretty_space(), b.clone()]);
    assert_layout(expected, actual);

    let actual = fuse_list(
        Some(sep.clone()),
        None,
        vec![a.clone(), b.clone(), c.clone()],
    );
    let expected = LayoutNode::concat(vec![
        a.clone(),
        sep.clone(),
        pretty_space(),
        b.clone(),
        sep.clone(),
        pretty_space(),
        c.clone(),
    ]);
    assert_layout(expected, actual);

    let actual = fuse_list(
        None,
        Some((c.clone(), d.clone())),
        vec![a.clone(), b.clone()],
    );
    let expected = LayoutNode::concat(vec![
        c.clone(),
        a.clone(),
        pretty_space(),
        b.clone(),
        d.clone(),
    ]);
    assert_layout(expected, actual);

    let actual = fuse_list(
        Some(sep.clone()),
        Some((c.clone(), d.clone())),
        vec![a.clone(), b.clone()],
    );
    let expected = LayoutNode::concat(vec![
        c.clone(),
        a.clone(),
        sep.clone(),
        pretty_space(),
        b.clone(),
        d.clone(),
    ]);
    assert_layout(expected, actual);
}
